src/chain_sync/network_context.rs

```diff
@@ -214,16 +246,14 @@ where
 
     /// Helper function to handle the peer retrieval if no peer supplied as well
     /// as the logging and updating of the peer info in the `PeerManager`.
-    async fn handle_chain_exchange_request<T>(
-        &self,
+    pub async fn handle_chain_exchange_request(
+        peer_manager: Arc<PeerManager>,
+        network_send: flume::Sender<NetworkMessage>,
         peer_id: Option<PeerId>,
         tsk: &TipsetKeys,
         request_len: u64,
         options: u64,
-    ) -> Result<Vec<T>, String>
-    where
-        T: TryFrom<TipsetBundle, Error = String> + Send + Sync + 'static,
-    {
+    ) -> Result<Vec<TipsetBundle>, String> {
```

Comment:
Looks like you're changing the semantics here. What happens if we are sent responses that can't be converted into full tipsets? Was the previous behavior wrong?

Reply:
The conversion now happens on the caller side, the logic should be same, see

https://github.com/ChainSafe/forest/pull/3593/files#diff-1686eeeaa959ad579a0b8ef0af0a7907816520a1c203b1077b12cda20444492bR174

https://github.com/ChainSafe/forest/pull/3593/files#diff-1686eeeaa959ad579a0b8ef0af0a7907816520a1c203b1077b12cda20444492bR196

Reply:
You've changed the semantics, though. With `fts` being a vector of results, the core logic has changed.

Reply:
I see what you mean, changed to `try_collect`.

(although the logic guarantees `len` of the Vec is 0 or 1, `Result<Vec<_>>` and `Vec<Result<_>>` makes no difference here.)